### PR TITLE
WT-7208 Leave table cursor in a valid state when subordinate index fails to open

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
@sueloverso, could you take a look at this please?

File: src/cursor/cur_table.c [hunk 1]
```diff
@@ -918,9 +919,24 @@ __curtable_open_indices(WT_CURSOR_TABLE *ctable)
...
-    return (0);
+
+    if (0) {
+err:
+        /*
+         * On failure, we can't leave a subset of the indices open, since the table cursor is still
+         * open. It's all or nothing, so we need to close them all, and leave things as they were
+         * before the first cursor operation.
+         */
+        for (i = 0, cp = ctable->idx_cursors; i < ctable->table->nindices; i++, cp++)
```
Comment (sueloverso) [inline src/cursor/cur_table.c:932]: Minor nit: you can just use `table->nindices` here.

File: src/cursor/cur_table.c [hunk 2]
```diff
@@ -918,9 +919,24 @@ __curtable_open_indices(WT_CURSOR_TABLE *ctable)
...
     WT_RET(__wt_calloc_def(session, table->nindices, &ctable->idx_cursors));
     for (i = 0, cp = ctable->idx_cursors; i < table->nindices; i++, cp++)
-        WT_RET(
+        WT_ERR(
           __wt_open_cursor(session, table->indices[i]->source, &ctable->iface, ctable->cfg, cp));
-    return (0);
+
+    if (0) {
+err:
+        /*
```
Comment (sueloverso) [inline src/cursor/cur_table.c:927]: This is all good. However you need to add a similar handling for colgroups in the function right above this one. And please add a colgroup test case to your new test.

**Review by sueloverso — APPROVED** [review]
LGTM but there is another case that needs the same fix.

File: src/cursor/cur_table.c (hunk 2 above)
Comment (ddanderson) [reply src/cursor/cur_table.c:927]: It turns out that colgroups are treated differently, they are opened when the table cursor is opened.  If column group open fails, the cursor open fails, so you can't get a "partially opened" cursor.

**Comment by ddanderson** [comment]
@sueloverso, can you take another look?   colgroups don't have the same problem, I added a longer comment to hopefully explain.  Please see if it makes sense.

**Review by sueloverso — APPROVED** [review]
Thanks for adding the comment. LGTM.

**Pull request merged**